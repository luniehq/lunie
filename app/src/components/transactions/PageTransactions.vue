<template>
  <TmPage
    data-title="Transactions"
    :loading="$apollo.queries.transactions.loading"
    :empty="dataEmpty"
    :empty-title="`No Transaction History`"
    :empty-subtitle="`There are no transactions associated with this address yet.`"
    :sign-in-required="true"
  >
    <template slot="signInRequired">
      <template>
        <EventList
          :events="transactions"
          :more-available="moreAvailable"
          @loadMore="loadMore"
        >
          <template slot-scope="event">
            <TransactionItem
              :key="event.key"
              :transaction="event"
              :validators="validatorsAddressMap"
              :address="address"
            />
          </template>
        </EventList>

        <TmDataMsg
          v-if="!$apollo.queries.transactions.loading"
          icon="calendar_today"
        >
          <div slot="title">Looking for older transactions?</div>
          <div slot="subtitle">
            <p>
              Lunie cannot display transactions from previous chains in your
              activity page.
            </p>
            <p>
              If you would like to view information from previous chain upgrades
              please visit our
              <a
                href="https://intercom.help/lunie/en/articles/3787014-how-to-get-blockchain-data-from-previous-chain-upgrades"
                >Help Center.</a
              >
            </p>
          </div>
        </TmDataMsg>
      </template>
    </template>
  </TmPage>
</template>

<script>
import { mapGetters } from "vuex"
import TmDataMsg from "common/TmDataMsg"
import TmPage from "common/TmPage"
import EventList from "common/EventList"
import TransactionItem from "transactions/TransactionItem"
import gql from "graphql-tag"
import uniqBy from "lodash.uniqby"

const txFields = `
  id
  type
  hash
  key
  height
  timestamp
  memo
  success
  fees {
    denom
    amount
  }
  details {
    ... on SendTx {
      from
      to
      amount {
        denom
        amount
      }
    }
    ... on StakeTx {
      to
      amount {
        denom
        amount
      }
    }
    ... on RestakeTx {
      to
      from
      amount {
        denom
        amount
      }
    }
    ... on UnstakeTx {
      from
      amount {
        denom
        amount
      }
    }
    ... on ClaimRewardsTx {
      from
      amounts {
        denom
        amount
      }
    }
    ... on SubmitProposalTx {
      proposalType
      proposalTitle
      proposalDescription
      initialDeposit {
        denom
        amount
      }
    }
    ... on VoteTx {
      proposalId
      voteOption
    }
    ... on DepositTx {
      proposalId
      amount {
        denom
        amount
      }
    }
  }
`

export default {
  name: `page-transactions`,
  components: {
    TransactionItem,
    EventList,
    TmDataMsg,
    TmPage,
  },
  data: () => ({
    pageNumber: 0,
    validators: [],
    transactions: [],
    dataLoaded: false,
    moreAvailable: true,
  }),
  computed: {
    ...mapGetters([`address`, `network`]),
    validatorsAddressMap() {
      const names = {}
      this.validators.forEach((item) => {
        names[item.operatorAddress] = item
      })
      return names
    },
    dataEmpty() {
      return this.transactions.length === 0
    },
  },
  methods: {
    loadMore() {
      // preload next transactions before scroll end and check if last loading loads new records
      // to prevent multiple requests
      if (this.dataLoaded === true) {
        // loads new portion
        this.pageNumber++
        this.dataLoaded = false
        this.$apollo.queries.transactions.fetchMore({
          // New variables
          variables: {
            networkId: this.network,
            address: this.address,
            pageNumber: this.pageNumber,
          },
          // Transform the previous result with new data
          updateQuery: (previousResult, { fetchMoreResult }) => {
            this.moreAvailable = fetchMoreResult.transactionsV2.length > 0
            return {
              // DEPRECATE uniqBy, should be resolved via API
              transactionsV2: uniqBy(
                [
                  ...previousResult.transactionsV2,
                  ...fetchMoreResult.transactionsV2,
                ],
                "key"
              ),
            }
          },
        })
      }
    },
  },
  apollo: {
    transactions: {
      query: gql`
        query transactionsV2($networkId: String!, $address: String!, $pageNumber: Int) {
          transactionsV2(networkId: $networkId, address: $address, pageNumber: $pageNumber) {
            ${txFields}
          }
        }
      `,
      skip() {
        return !this.address
      },
      variables() {
        return {
          networkId: this.network,
          address: this.address,
          pageNumber: 0,
        }
      },
      update(result) {
        this.dataLoaded = true
        this.moreAvailable = result.transactionsV2.length > 0
        return result.transactionsV2
      },
      subscribeToMore: {
        document: gql`
          subscription($networkId: String!, $address: String!) {
            userTransactionAddedV2(networkId: $networkId, address: $address) {
              ${txFields}
            }
          }
        `,
        updateQuery: (previousResult, { subscriptionData }) => {
          if (previousResult && subscriptionData.data.userTransactionAddedV2) {
            return {
              transactionsV2: [
                subscriptionData.data.userTransactionAddedV2,
                ...previousResult.transactionsV2,
              ],
            }
          }
        },
        /* istanbul ignore next */
        variables() {
          return {
            networkId: this.network,
            address: this.address,
          }
        },
      },
    },
    validators: {
      query: gql`
        query validators($networkId: String!) {
          validators(networkId: $networkId) {
            id
            name
            operatorAddress
            picture
          }
        }
      `,
      skip() {
        return !this.address
      },
      variables() {
        return {
          networkId: this.network,
        }
      },
    },
  },
}
</script>
