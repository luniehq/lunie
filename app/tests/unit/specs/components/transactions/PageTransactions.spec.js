import PageTransactions from "transactions/PageTransactions"
import { createLocalVue, shallowMount } from "@vue/test-utils"

describe(`PageTransactions`, () => {
  const localVue = createLocalVue()
  localVue.directive(`tooltip`, () => {})

  const addresses = [
    `cosmos15ky9du8a2wlstz6fpx3p4mqpjyrm5ctpesxxn9`,
    `cosmos1pxdf0lvq5jvl9uxznklgc5gxuwzpdy5ynem546`,
  ]

  const validators = [
    { name: "cosmos1a", operatorAddress: "cosmos1a" },
    { name: "cosmos1b", operatorAddress: "cosmos1b" },
    { name: "cosmos1c", operatorAddress: "cosmos1c" },
  ]

  const transactions = [
    {
      type: "cosmos-sdk/MsgUndelegate",
      value: {
        delegator_address: "cosmos3",
        validator_address: "cosmos1e",
        amount: { denom: "uatom", amount: "50000" },
      },
      key:
        'cosmos-sdk/MsgUndelegate_2019-07-31T09:22:23.054Z_{"delegator_address":"cosmos1jq9mc3kp4nnxwryr09fpqjtrwya8q5q480zu0e","validator_address":"cosmos1a","amount":{"denom":"uatom","amount":"50000"}}',
      height: 1248479,
      timestamp: "2019-07-31T09:22:23.054Z",
      memo: "",
      fee: { amount: "4141", denom: "ATOM" },
      group: "staking",
    },
    {
      type: "cosmos-sdk/MsgSubmitProposal",
      value: {
        proposer: "cosmos15ky9du8a2wlstz6fpx3p4mqpjyrm5ctpesxxn9",
        proposal_type: "Text",
        title: "Test Proposal",
        description: "This is a test proposal",
        initial_deposit: [{ denom: "STAKE", amount: "100" }],
      },
      key:
        'cosmos-sdk/MsgSubmitProposal_undefined_{"proposer":"cosmos15ky9du8a2wlstz6fpx3p4mqpjyrm5ctpesxxn9","proposal_type":"Text","title":"Test Proposal","description":"This is a test proposal","initial_deposit":[{"denom":"STAKE","amount":"100"}]}',
      height: 56673,
      timestamp: null,
      fee: { amount: "0", denom: "ATOM" },
      group: "governance",
      liquidDate: null,
    },
    {
      type: "cosmos-sdk/MsgDeposit",
      value: {
        depositor: "cosmos15ky9du8a2wlstz6fpx3p4mqpjyrm5ctpesxxn9",
        proposal_id: "1",
        amount: [{ denom: "STAKE", amount: "100" }],
      },
      key:
        'cosmos-sdk/MsgDeposit_undefined_{"depositor":"cosmos15ky9du8a2wlstz6fpx3p4mqpjyrm5ctpesxxn9","proposal_id":"1","amount":[{"denom":"STAKE","amount":"100"}]}',
      height: 213,
      timestamp: null,
      fee: { amount: "0", denom: "ATOM" },
      group: "governance",
      liquidDate: null,
    },
    {
      type: "cosmos-sdk/BeginUnbonding",
      value: {
        validator_address: "cosmos1a",
        delegator_address: "cosmos15ky9du8a2wlstz6fpx3p4mqpjyrm5ctpesxxn9",
        shares: "5",
      },
      key:
        'cosmos-sdk/BeginUnbonding_undefined_{"validator_address":"cosmos1a","delegator_address":"cosmos15ky9du8a2wlstz6fpx3p4mqpjyrm5ctpesxxn9","shares":"5"}',
      height: 170,
      timestamp: null,
      fee: { amount: "0", denom: "ATOM" },
      liquidDate: null,
    },
  ]

  let wrapper, $store, $apollo

  const state = {
    session: {
      address: addresses[0],
      signedIn: true,
    },
    connection: {
      network: "awesomenet",
    },
  }

  $apollo = {
    queries: {
      validators: {
        loading: false,
      },
      transactions: {
        loading: false,
        variables: jest.fn(),
        fetchMore: jest.fn(() => ({
          variables: jest.fn(),
          updateQuery: jest.fn(),
        })),
      },
    },
  }

  beforeEach(() => {
    $store = {
      state,
      getters: {
        address: "cosmos1",
        network: "cosmos-hub-mainnet",
      },
    }

    wrapper = shallowMount(PageTransactions, {
      localVue,
      mocks: {
        $store,
        $apollo,
      },
      propsData: {
        transactions,
        validators,
      },
    })
    wrapper.setData({ transactions, validators })
  })

  describe(`Renders correctly`, () => {
    it(`and loads transactions`, async () => {
      expect(wrapper.element).toMatchSnapshot()
    })

    it(`and does not load transactions if the user has not signed in`, async () => {
      $store.state.session.signedIn = false
      $store.state.session.address = undefined

      wrapper = shallowMount(PageTransactions, {
        localVue,
        mocks: {
          $store,
          $apollo,
        },
        directives: {
          infiniteScroll: () => {},
        },
      })
      wrapper.setProps({ transactions, validators })
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  it(`should refresh the transaction history when signed in`, async () => {
    $store.state.session.signedIn = false
    $store.state.session.address = undefined

    wrapper = shallowMount(PageTransactions, {
      localVue,
      mocks: {
        $store,
        $apollo,
      },
      directives: {
        infiniteScroll: () => {},
      },
    })
    $store.state.session.signedIn = true
    $store.state.session.address = undefined
  })

  it("transactions updated on subscription trigger", () => {
    const self = {
      loadedTransactions: [],
    }
    let newTransaction = {
      type: "cosmos-sdk/MsgUndelegate",
      value: JSON.stringify({
        delegator_address: "cosmos2",
        validator_address: "cosmos4de",
        amount: { denom: "uatom", amount: "10000" },
      }),
      key:
        'cosmos-sdk/MsgUndelegate_2019-07-31T09:22:23.054Z_{"delegator_address":"cosmos1jq9mc3kp4nnxwryr09fpqjtrwya8q5q480zu0e","validator_address":"cosmos1a","amount":{"denom":"uatom","amount":"50000"}}',
      height: 1248479,
      timestamp: "2019-07-31T09:22:23.054Z",
      memo: "",
      fee: { amount: "4141", denom: "ATOM" },
      group: "staking",
    }
    let result = PageTransactions.apollo.transactions.update.call(self, {
      transactionsV2: [newTransaction],
    })
    // adjusting result
    newTransaction.value = JSON.parse(newTransaction.value)
    newTransaction.timestamp = new Date(newTransaction.timestamp)
    expect(result).toEqual([newTransaction])
  })

  it(`should load more transactions on loadMore action`, async () => {
    wrapper.setData({
      dataLoaded: true,
    })
    wrapper.vm.loadMore()
    // pageNumber should be updated
    expect(wrapper.vm.pageNumber).toBeGreaterThan(0)
  })

  it("transaction added on subscription trigger", () => {
    const self = {
      transactions: [],
    }
    let result = PageTransactions.apollo.transactions.subscribeToMore.updateQuery.call(
      self,
      {
        transactionsV2: [],
      },
      {
        subscriptionData: {
          data: {
            userTransactionAddedV2: {
              type: "cosmos-sdk/MsgUndelegate",
              value: {
                delegator_address: "cosmos2",
                validator_address: "cosmos4de",
                amount: { denom: "uatom", amount: "10000" },
              },
              height: 1248479,
              timestamp: "2019-07-31T09:22:23.054Z",
              memo: "",
              fee: { amount: "4141", denom: "ATOM" },
            },
          },
        },
      }
    )

    expect(result.transactionsV2.length).toBeGreaterThan(0)
  })

  it(`should not load more if currently loading`, async () => {
    wrapper = shallowMount(PageTransactions, {
      localVue,
      mocks: {
        $store,
        $apollo,
      },
      directives: {
        infiniteScroll: () => {},
      },
    })
    wrapper.setData({
      dataLoaded: false,
    })
    wrapper.vm.loadMore()
    // pageNumber should not have updated
    expect(wrapper.vm.pageNumber).toBe(0)
  })

  it(`validator address map to be correct`, async () => {
    wrapper = shallowMount(PageTransactions, {
      localVue,
      mocks: {
        $store,
        $apollo,
      },
      directives: {
        infiniteScroll: () => {},
      },
    })
    wrapper.setData({ validators })
    expect(Object.keys(wrapper.vm.validatorsAddressMap)).toEqual([
      "cosmos1a",
      "cosmos1b",
      "cosmos1c",
    ])
  })
})
