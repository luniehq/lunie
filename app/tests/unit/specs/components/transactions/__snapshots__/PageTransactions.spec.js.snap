// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageTransactions Renders correctly and does not load transactions if the user has not signed in 1`] = `
<tmpage-stub
  data-title="Transactions"
  empty="true"
  emptysubtitle="There are no transactions associated with this address yet."
  emptytitle="No Transaction History"
  loaderpath=""
  signinrequired="true"
>
  <template>
    <eventlist-stub
      events=""
      moreavailable="true"
    />
     
    <tmdatamsg-stub
      icon="calendar_today"
      iconcolor="var(--warning)"
    >
      <div>
        Looking for older transactions?
      </div>
       
      <div>
        <p>
          
            Lunie cannot display transactions from previous chains in your
            activity page.
          
        </p>
         
        <p>
          
            If you would like to view information from previous chain upgrades
            please visit our
            
          <a
            href="https://intercom.help/lunie/en/articles/3787014-how-to-get-blockchain-data-from-previous-chain-upgrades"
          >
            Help Center.
          </a>
        </p>
      </div>
    </tmdatamsg-stub>
  </template>
</tmpage-stub>
`;

exports[`PageTransactions Renders correctly and loads transactions 1`] = `
<tmpage-stub
  data-title="Transactions"
  emptysubtitle="There are no transactions associated with this address yet."
  emptytitle="No Transaction History"
  loaderpath=""
  signinrequired="true"
  transactions="[object Object],[object Object],[object Object],[object Object]"
  validators="[object Object],[object Object],[object Object]"
>
  <template>
    <eventlist-stub
      events="[object Object],[object Object],[object Object],[object Object]"
      moreavailable="true"
    />
     
    <tmdatamsg-stub
      icon="calendar_today"
      iconcolor="var(--warning)"
    >
      <div>
        Looking for older transactions?
      </div>
       
      <div>
        <p>
          
            Lunie cannot display transactions from previous chains in your
            activity page.
          
        </p>
         
        <p>
          
            If you would like to view information from previous chain upgrades
            please visit our
            
          <a
            href="https://intercom.help/lunie/en/articles/3787014-how-to-get-blockchain-data-from-previous-chain-upgrades"
          >
            Help Center.
          </a>
        </p>
      </div>
    </tmdatamsg-stub>
  </template>
</tmpage-stub>
`;
